[jit] Added a new operator for concat that takes in variadic parameters

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59881 [jit] Added a pass to transform aten::cat ops to prim::Concat op with variable number of inputs
* **#59880 [jit] Added a new operator for concat that takes in variadic parameters**

Differential Revision: [D29277877](https://our.internmc.facebook.com/intern/diff/D29277877)

navahgar committed:
[jit] Added a new operator for concat that takes in variadic parameters

[ghstack-poisoned]

eellison commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -745,6 +745,19 @@ RegisterOperators reg(
          },
          aliasAnalysisSpecialCase()),
+     OperatorGenerator(
+         TORCH_SELECTIVE_SCHEMA("prim::Concat(...) -> Tensor"),
```
So, the convention here is that it's (Tensors..., dim) right ?

navahgar replied:
Yes. Added a comment above.

eellison commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -745,6 +745,19 @@ RegisterOperators reg(
+           for (int i = 0; i < num_inputs - 1; ++i) {
+             inputs[i] = pop(stack).toTensor();
+           }
+           std::reverse(inputs.begin(), inputs.end());
```
would there be a way of writing this so we don't need to have the reverse ?

eellison commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -745,6 +745,19 @@ RegisterOperators reg(
          },
          aliasAnalysisSpecialCase()),
+     OperatorGenerator(
+         TORCH_SELECTIVE_SCHEMA("prim::Concat(...) -> Tensor"),
```
I think we probably want `prim::ConstantConcat`, following along the lines of `prim::ConstantChunk`. I'm not sure though.

navahgar replied:
I'm open to changing that, if everyone prefers that.

eellison reviewed (commented):
LGTM - this PR doesn't turn it on but i'd be curious to hear from @Krovatkin @ZolotukhinM @bertmaher on their opinions here. 

Also cc @haolu, should we get this turned on in static runtime? might reduce overhead of constructing input list

navahgar committed:
Update on "[jit] Added a new operator for concat that takes in variadic parameters"




[ghstack-poisoned]

navahgar committed:
Update on "[jit] Added a new operator for concat that takes in variadic parameters"




[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29277877).

navahgar committed:
Update on "[jit] Added a new operator for concat that takes in variadic parameters"


Differential Revision: [D29277877](https://our.internmc.facebook.com/intern/diff/D29277877)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29277877).

navahgar committed:
Update on "[jit] Added a new operator for concat that takes in variadic parameters"


Differential Revision: [D29277877](https://our.internmc.facebook.com/intern/diff/D29277877)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29277877).

navahgar committed:
Update on "[jit] Added a new operator for concat that takes in variadic parameters"


Differential Revision: [D29277877](https://our.internmc.facebook.com/intern/diff/D29277877)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29277877).
